### DISCUSSION TITLE
Problem with Widget UiItemsProveder class

### DISCUSSION THREAD

**Discussion by sagilar** [iTwin/itwinjs-core#3300] [Q&A] [answered]
Hi,

I'm trying to implement an example based on some of the samples and a [tutorial](https://developer.bentley.com/tutorials/itwin-viewer-hello-world/), but I get an error when trying to get the widget to the viewer.
I don't know if the problem is due to the versions (the samples say they are for iTwin version 2.x and I'm running version 3.0). Anyway, do you know how to solve the problem? I attach here the code that I'm using from the samples.

**Widget:**  
```
export class Simple3dWidgetProvider implements UiItemsProvider {
  public readonly id: string = "Simple3dWidgetProvider";

  public provideWidgets(_stageId: string, _stageUsage: string, location: StagePanelLocation, _section?: StagePanelSection): ReadonlyArray<AbstractWidgetProps> {
    const widgets: AbstractWidgetProps[] = [];
    if (location === StagePanelLocation.Right) {
      widgets.push(
        {
          id: "Simple3dWidget",
          label: "Simple 3d Controls",
          defaultState: WidgetState.Floating,
          getWidgetContent: () => {
            return <Simple3dWidget />;
          },
        }
      );
    }
    return widgets;
  }  
}
```

  
**App main:**  
```
const App: React.FC = () => {  
/* Content of the function here */  
  return (
    <div className="viewer-container">
      <Viewer
        theme={"light"}
        uiProviders={uiProviders}
        authClient={authClient}
        viewCreatorOptions={viewCreatorOptions}
        enablePerformanceMonitors={true}
      />
    </div>
  );
};
export default App;
```
  
And the error that I get: `Type 'Simple3dWidgetProvider[]' is not assignable to type 'UiItemsProvider[]'.
  Type 'Simple3dWidgetProvider' is not assignable to type 'UiItemsProvider'.
    The types returned by 'provideWidgets(...)' are incompatible between these types.` at line **`uiProviders={uiProviders}`**.  

Thanks in advance!

**Comment by aruniverse** [comment, marked as answer]
The Viewer's uiProvider prop takes an array of UiItemsProviders. Please show the code for what you're passing in. What does your uiPoviders variable look like?

**Reply by sagilar** [reply]
I just followed the tutorials, so I have the code for the uiProviders looks this way:  

```
import { Simple3dWidgetProvider } from "./Simple3dWidget";


const uiProviders = [new Simple3dWidgetProvider()];
```

**Reply by BaoTon** [reply]
Hi @sagilar , I think you may be missing another "return widgets" under the 2nd to last bracket. You're only returning widgets if the condition matches, provideWidgets expects you to return AbstractWidgetProps.

**Reply by BaoTon** [reply]
Nevermind... I don't know the issue. I copied and pasted your exact code using my own component and it worked fine.

**Reply by BaoTon** [reply]
@sagilar , could you verify where import UiItemsProvider is coming from at the top of the file?

`import { UiItemsProvider } from "@itwin/appui-abstract";`

It should be from @itwin/appui-abstract

**Reply by sagilar** [reply]
I changed the libraries from \@bentley to \@itwin and it partially worked. I'm trying to follow the sample of [Simple 3D](https://www.itwinjs.org/sandboxes/iTwinPlatform/Simple%203d), but it seems that it has several inconsistencies because of the outdated version of the sample, so in the end, I can't run it. Could you provide any updated example for the current iTwin version (3.x)?

**Reply by aruniverse** [reply]
Could you elaborate on the "several inconsistencies" you're seeing? 

We have a test viewer for development in the [viewer repo](https://github.com/iTwin/viewer/tree/master/packages/apps/web-viewer-test), hopefully that can help

**Reply by NancyMcCallB** [reply]
Hi, @sagilar . Here's a 3.0 sandbox that provides a widget to the viewer. I don't see the error in your code, but comparing to this might help you find the problem.
https://www.itwinjs.org/sandboxes/NancyMcCall/Hide%20and%20Show%20Widget%20as%20Desired

**Reply by NancyMcCallB** [reply]
@sagilar try this: 
const uiProviders: UiItemsProvider[] = [new Simple3dWidgetProvider()];

**Reply by sagilar** [reply]
> Hi, @sagilar . Here's a 3.0 sandbox that provides a widget to the viewer. I don't see the error in your code, but comparing to this might help you find the problem. https://www.itwinjs.org/sandboxes/NancyMcCall/Hide%20and%20Show%20Widget%20as%20Desired

I was trying to run this, but this time the problem seems to be with the package `@itwin/sandbox` that I can't install.

**Reply by NancyMcCallB** [reply]
Hi, 

> > Hi, @sagilar . Here's a 3.0 sandbox that provides a widget to the viewer. I don't see the error in your code, but comparing to this might help you find the problem. https://www.itwinjs.org/sandboxes/NancyMcCall/Hide%20and%20Show%20Widget%20as%20Desired
> 
> I was trying to run this, but this time the problem seems to be with the package `@itwin/sandbox` that I can't install.

The sandbox package does the setup for the sandboxes themselves, so you can't use it in your app.

Did you try explicitly typing your uiProviders variable as I suggested above?

**Reply by sagilar** [reply]
I realized you updated all the sandboxes to iTwin version 3.x, so I took one of the examples, and now is running. It seemed to be a problem between using a code for version 2.x and having version 3.x of the library.
Thanks for the support!
